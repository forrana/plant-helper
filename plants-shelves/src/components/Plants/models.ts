export interface PlantType {
  id: number;
  name: string;
  scientificName: string;
  daysUntilNextWatering: number;
  daysBetweenWatering: number;
  daysBetweenWateringGrowing: number;
  daysBetweenWateringDormant: number;
  symbol: SymbolType;
  daysPostpone: number;
}

export interface SymbolType {
  userWideId: number
}

export interface PlantsData {
  plants: PlantType[];
}

export interface Subscription {
  endpoint: string,
  p256dh: string,
  auth: string,
  permission_given: boolean
}

export interface SubscriptionData {
  subscription: boolean
}

export interface PlantData {
  plant: PlantType;
}

export interface PlantSuggestion {
  nickName: string,
  nickNames: Array<string>,
  scientificName: string,
  daysBetweenWateringGrowingInt: number,
  daysBetweenWateringDormantInt: number
}

export interface PlantNickName {
  name: string,
  plantEntry: PlantSuggestion
}

export interface GlobalState extends PlantsData { }

export interface loadAction { type: 'load', plants: PlantType[] }
export interface addAction { type: 'add', plant: PlantType }
export interface updateAction { type: 'update', plant: PlantType, index: number }
export interface deleteAction { type: 'delete', index: number }

export type GlobalReducerAction =
  | addAction
  | loadAction
  | updateAction
  | deleteAction;
