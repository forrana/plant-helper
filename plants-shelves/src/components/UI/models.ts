export interface MessageType { description: string, color: "danger" | "info" | "warning" | "success" }

export interface AddMessage  { type: 'addMessage', message: MessageType }
export interface RemoveMessage  { type: 'removeMessage' }

export interface AlertState { messages: MessageType[] }

export type AlertReducerAction =
  | AddMessage
  | RemoveMessage