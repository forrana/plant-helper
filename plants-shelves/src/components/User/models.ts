export interface LoginError {
  message: string;
  code: string;
}
