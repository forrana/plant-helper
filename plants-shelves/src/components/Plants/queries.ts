import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
    symbol {
      userWideId
    }
    room {
      id,
      roomName,
      colorBackground
    }
  }
}
`;

const GET_ALL_ROOMS = gql`
  query {
    rooms {
      id, roomName, colorBackground
    }
  }
`;
const ROOMS_BY_NAME_FRAGMENT = gql`
  query RoomsByNameFragment($nameFragment: String!) {
    roomsByNameFragment(nameFragment: $nameFragment) {
      id, roomName, colorBackground
    }
  }
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $groupName: String, $colorBackground: String) {
      createPlant(plantName: $plantName, scientificName: $scientificName, daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
        daysBetweenWateringDormant: $daysBetweenWateringDormant, groupName: $groupName, colorBackground: $colorBackground) {
        plant {
          id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
          symbol {
            userWideId
          }
          room {
            id,
            roomName,
            colorBackground
          }
        }
      }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const POSTPONE_WATERING = gql`
  mutation PostponeWatering($plantId: ID!, $postponeDays: Int!) {
    postponeWatering(plantId: $plantId, days: $postponeDays) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant, daysPostpone
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_PLANT = gql`
  mutation UpdatePlant($plantId: ID!, $plantName: String!,
    $scientificName: String!, $daysBetweenWateringGrowing: Int!, $daysBetweenWateringDormant: Int!,
    $postponeDays: Int!, $groupName: String, $colorBackground: String) {
    updatePlant(plantId: $plantId, plantName: $plantName, scientificName: $scientificName,
      daysBetweenWateringGrowing: $daysBetweenWateringGrowing,
      daysBetweenWateringDormant: $daysBetweenWateringDormant,
      postponeDays: $postponeDays,
      groupName: $groupName,
      colorBackground: $colorBackground) {
      plant {
        id, name, scientificName, daysUntilNextWatering,
        daysBetweenWatering, daysBetweenWateringGrowing, daysBetweenWateringDormant,
        daysPostpone,
        symbol {
          userWideId
        }
        room {
          id,
          roomName,
          colorBackground
        }
      }
    }
  }
`;

const UPDATE_ROOM = gql`
  mutation UpdateRoom($roomId: ID!, $roomName: String!) {
    updateRoom(roomId: $roomId, roomName: $roomName) {
      ok
      room {
        id,
        roomName,
        colorBackground
      }
    }
  }
`

const CREATE_ROOM = gql`
  mutation CreateRoom($plant1Id: ID!, $plant2Id: ID!) {
    createRoom(plant1Id: $plant1Id, plant2Id: $plant2Id) {
      ok
    }
  }
`;

const DELETE_PLANT = gql`
  mutation DeletePlant($plantId: ID!) {
    deletePlant(plantId: $plantId) {
      ok
    }
  }
`;

const GET_SUBSCRIPTION = gql`
  query {
    subscription
  }
`

const CREATE_SUBSCRIPTION = gql`
  mutation CreateSubscription($endpoint: String!, $p256dh: String!, $auth: String!, $permissionGiven: Boolean!) {
    createSubscription(endpoint: $endpoint, p256dh: $p256dh, auth: $auth, permissionGiven: $permissionGiven) {
      ok
    }
  }
`

export {
  GET_ALL_PLANTS,
  GET_ALL_ROOMS,
  ADD_PLANT,
  WATER_PLANT,
  UPDATE_PLANT,
  DELETE_PLANT,
  CREATE_SUBSCRIPTION,
  GET_SUBSCRIPTION,
  POSTPONE_WATERING,
  CREATE_ROOM,
  ROOMS_BY_NAME_FRAGMENT,
  UPDATE_ROOM
}
