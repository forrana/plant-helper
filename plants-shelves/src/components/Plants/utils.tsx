import { useEffect, useState } from "react";
import { JWTToken } from "../User/models";

const getRandomString = (string_length: number): string => [...Array(string_length)].map(i=>(~~(Math.random()*36)).toString(36)).join('')

function useDebounce(value: string, delay: number) {
    // State and setters for debounced value
    const [debouncedValue, setDebouncedValue] = useState(value);
    useEffect(
      () => {
        // Update debounced value after delay
        const handler = setTimeout(() => {
          setDebouncedValue(value);
        }, delay);
        // Cancel the timeout if value changes (also on delay change or unmount)
        // This is how we prevent debounced value from updating if value is changed ...
        // .. within the delay period. Timeout gets cleared and restarted.
        return () => {
          clearTimeout(handler);
        };
      },
      [value, delay] // Only re-call effect if value or delay changes
    );
    return debouncedValue;
  }

const generateColorForGroup = (groupName: string) => {
  // @ts-ignore: iterate through string is needed
  let groupNumber = [...groupName].reduce((prev, current) => prev + current.charCodeAt(), 0)
  let randomColor = "#" + Math.floor(Math.random()*(16777215 - groupNumber)).toString(16);
  return randomColor;
}

const parseJwt = (token: string, separator: string = ".", position: number = 1): JWTToken => {
  try {
    return JSON.parse(atob(token.split(separator)[position]));
  } catch (e) {
    throw new Error("JWTToken parsing error");
  }
};

export { getRandomString, useDebounce, generateColorForGroup, parseJwt }