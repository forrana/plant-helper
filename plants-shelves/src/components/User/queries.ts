import { gql } from '@apollo/client';

const CREATE_USER = gql`
  mutation Register($email: String!, $username: String!, $password1: String!, $password2: String!) {
    register(email: $email, username: $username, password1: $password1, password2: $password2) {
      success,
      errors
    }
  }
`;

const LOG_IN = gql`
  mutation TokenAuth($username: String!, $password: String!) {
    tokenAuth(username: $username, password: $password) {
      success,
      errors,
      unarchiving,
      token,
      refreshToken,
      unarchiving,
      user {
        id,
        username,
      }
    }
  }
`;


const REFRESH_TOKEN = gql`
  mutation RefreshToken($refreshToken: String!) {
    refreshToken(
      refreshToken: $refreshToken
    ) {
      success,
      errors,
      payload,
      token,
      refreshToken,
    }
  }
`

const UPSERT_USER_SETTINGS = gql`
  mutation UpsertUserSettings($startTime: String!, $endTime: String!, $timezone: String) {
    upsertUserSettings(startTime: $startTime, endTime: $endTime, timezone: $timezone) {
      ok
    }
  }
`

const GET_USER_SETTINGS = gql`
  query {
    userSettings {
      notificationsStartTime, notificationsEndTime, timezone
    }
  }
`;

export { CREATE_USER, LOG_IN, REFRESH_TOKEN, UPSERT_USER_SETTINGS, GET_USER_SETTINGS }
