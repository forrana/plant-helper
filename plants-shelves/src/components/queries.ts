import { gql } from '@apollo/client';
const GET_ALL_PLANTS = gql`
query {
  plants {
    id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
  }
}
`;

const ADD_PLANT = gql`
  mutation CreatePlant($plantName: String!, $scientificName: String!) {
    createPlant(plantName: $plantName, scientificName: $scientificName) {
      plant {
        name,
        scientificName
      }
    }
  }
`;

const WATER_PLANT = gql`
  mutation WaterPlant($plantId: ID!) {
    waterPlant(plantId: $plantId) {
      plant {
        id, name, scientificName, daysUntilNextWatering, daysBetweenWatering
      }
    }
  }
`;


export { GET_ALL_PLANTS, ADD_PLANT, WATER_PLANT }
