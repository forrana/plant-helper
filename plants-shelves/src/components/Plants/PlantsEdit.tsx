import React, { useState, useContext } from 'react';
import { Redirect } from "react-router-dom";
import { useMutation } from '@apollo/client';
import { Button, Form, FormGroup, Input, InputGroup, InputGroupText, Label, Modal, ModalBody, ModalFooter, ModalHeader, Spinner } from 'reactstrap';

import { DELETE_PLANT, UPDATE_PLANT } from './queries'
import PlantsDispatch from './PlantsDispatch';
import { PlantData, PlantNickName, RoomType } from './models'
import styles from "./Plant.module.css"
import editStyles from "./PlantsEdit.module.css"
import uiStyles from "../UI/UIElements.module.css"
import ErrorHandler from './ErrorHandler';
import pot from './images/pot.png'
import PlantNameInput from './PlantNameInput';
import RoomNameInput from './RoomNameInput';

interface PlantsEditProps extends PlantData { index: number, action?: () => any }

function PlantsEdit({ plant, index, action }: PlantsEditProps) {
  const dispatch = useContext(PlantsDispatch);
  const [submitted, setSubmitted] = useState(false);
  const [plantName, setPlantName] = useState(plant.name);
  const [groupName, setGroupName] = useState(plant.room?.roomName || "");
  const [groupColor, setGroupColor] = useState("#FFFFFF")
  const [scientificName, setScientificName] = useState(plant.scientificName);
  const [daysBetweenWateringGrowing, setDaysBetweenWateringGrowing] = useState(plant.daysBetweenWateringGrowing)
  const [daysBetweenWateringDormant, setDaysBetweenWateringDormant] = useState(plant.daysBetweenWateringDormant)
  const [postponeDays, setPostponeDays] = useState(plant.daysPostpone)


  const [updatePlant, { loading, error }] = useMutation(UPDATE_PLANT, {
    onCompleted: (data: { updatePlant: PlantData }) => {
      dispatch && dispatch({ type: 'update', plant: data.updatePlant.plant, index: index});
      action && action();
      setSubmitted(true);
    },
    onError: (e) => console.error('Error creating plant:', e)
  });

  const generateColorForGroup = (groupName: string) => {
    // @ts-ignore: iterate through string is needed
    let groupNumber = [...groupName].reduce((prev, current) => prev + current.charCodeAt(), 0)
    let randomColor = "#" + Math.floor(Math.random()*(16777215 - groupNumber)).toString(16);
    return randomColor;
  }

  const setPlantSettings = (plantSuggestion: PlantNickName) => {
    setPlantName(plantSuggestion.name);
    setScientificName(plantSuggestion.plantEntry.scientificName);
    setDaysBetweenWateringGrowing(plantSuggestion.plantEntry.daysBetweenWateringGrowingInt);
    setDaysBetweenWateringDormant(plantSuggestion.plantEntry.daysBetweenWateringDormantInt);
  }

  const setRoomNameFromSuggestion = (roomSuggestion: RoomType) => {
    const newValue = roomSuggestion.roomName;
    setGroupColor(generateColorForGroup(newValue))
    setGroupName(newValue)
  }

  const handlePlantNameInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPlantName(event.target.value);
  };

  const handleGroupNameInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;
    setGroupColor(generateColorForGroup(newValue))
    setGroupName(newValue);
  };

  const handleScientificNameInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setScientificName(event.target.value);
  };

  const handleDaysBetweenWateringGrowingInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setDaysBetweenWateringGrowing(parseInt(event.target.value))
  }

  const handleDaysBetweenWateringDormantInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setDaysBetweenWateringDormant(parseInt(event.target.value))
  }


  const handleDaysPostponeInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPostponeDays(parseInt(event.target.value))
  }

  const handleFormSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    if(plantName && scientificName) {
      updatePlant({ variables: { plantId: plant.id, plantName, scientificName, daysBetweenWateringDormant, daysBetweenWateringGrowing, postponeDays, groupName } })
    }
  }

  const [modal, setModal] = useState(false);
  const toggleModal = () => setModal(!modal);

  const [deletePlant, deletingStatus] = useMutation(DELETE_PLANT, {
    onCompleted: () => {
      dispatch && dispatch({ type: 'delete', index: index })
    },
    onError: (e) => console.error('Error deleting plant:', e)
  });

  const confirmDeletion = () => {
    deletePlant({variables: { plantId: plant.id } });
    toggleModal()
  }

  if (submitted) return <Redirect push to="/"/>

  if (loading || deletingStatus.loading) return  <Spinner color="primary" />

  if (error) {
    return  <ErrorHandler error={error} />
  }

  if (deletingStatus.error) {
    return  <ErrorHandler error={deletingStatus.error} />
  }

  return (
    <>
    <Form
      onSubmit={handleFormSubmit}
      autoComplete="off"
      autoFocus={false}
    >
      <section className={styles.viewControls}>
        <span></span>
        <Button outline color="danger" onClick={toggleModal} title="Remove" data-testid="remove-btn">Delete!</Button>
      </section>
      <img src={pot} alt="plant pot" className={editStyles.image}/>
      <FormGroup>
        <Label for="name">Name:</Label>
        <PlantNameInput type="text" title="Plant name" name="name" id="name" data-testid="plant-name-input" placeholder="Plant name"
          value={plantName}
          setValue={setPlantSettings}
          onChange={handlePlantNameInputChange}
          autoFocus={true}
          required
        />
      </FormGroup>
      <FormGroup>
        <Label for="scientificName">Scientific name:</Label>
        <Input
          type="text" name="scientificName" id="scientificName" placeholder="Scientific name"
          data-testid="plant-scientific-name-input"
          value={scientificName}
          onChange={handleScientificNameInputChange}
          required
          />
      </FormGroup>
      <FormGroup>
        <Label for="scientificName">Group:</Label>
        <InputGroup className={editStyles.autoInputWithColorPicker}>
          <RoomNameInput type="text" name="groupName" id="groupName" placeholder="Group name"
            className={editStyles.groupInput}
            data-testid="plant-group-name-input"
            value={groupName}
            setValue={setRoomNameFromSuggestion}
            onChange={handleGroupNameInputChange}
          />
          <InputGroupText className={editStyles.colorPickerContainer}>
            <Input type="color" name="groupColor" placeholder="Group color" alt="Group color"
              className={editStyles.colorPickerInput}
              data-testid="plant-group-color-input"
              setValue={setGroupColor}
              value={groupColor}
            />
          </InputGroupText>
        </InputGroup>
      </FormGroup>
      <FormGroup>
        <Label for="postponeDays">Postpone days:</Label>
        <Input
          min={0} max={100} type="number" step="1"
          name="postponeDays" id="postponeDays" placeholder="Postpone days"
          data-testid="plant-days-postpone-input"
          value={postponeDays}
          setValue={setPostponeDays}
          onChange={handleDaysPostponeInputChange}
          required
          />
      </FormGroup>
      <FormGroup>
        <section className={styles.label}>
          <i className="icon icon-droplet"></i>
          <h4 className={styles.header}><i className="icon icon-droplet"></i></h4>
        </section>
        <Input type="range" name="daysBetweenWateringGrowing" id="daysBetweenWateringGrowing" placeholder="Days between watering growing"
            value={daysBetweenWateringGrowing}
            data-testid="days-between-watering-growing-input"
            onChange={handleDaysBetweenWateringGrowingInputChange}
            autoFocus={true}
            min={1}
            max={30}
            required
          />
        <div className="text-center"><small><b>Growing season</b> {daysBetweenWateringGrowing} day(s) between waterings</small></div>
      </FormGroup>
      <FormGroup>
        <section className={styles.label}>
          <i className="icon icon-droplet"></i>
          <h4 className={styles.header}><i className="icon icon-droplet"></i></h4>
        </section>
        <Input type="range" name="daysBetweenWateringDormant" id="daysBetweenWateringDormant" placeholder="Days between watering dormant"
            value={daysBetweenWateringDormant}
            data-testid="days-between-watering-dormant-input"
            onChange={handleDaysBetweenWateringDormantInputChange}
            autoFocus={true}
            min={1}
            max={30}
            required
          />
        <div className="text-center"><small><b>Dormant season</b> {daysBetweenWateringDormant} day(s) between waterings</small></div>
      </FormGroup>
      <section className={styles.viewControls}>
        <Button color="success" title="Save!" type="submit">Save changes!</Button>
        <Button outline color="danger" title="Cancel!" onClick={action}>Cancel</Button>
      </section>
    </Form>
    <Modal isOpen={modal} toggle={toggleModal} autoFocus={false}>
        <ModalHeader toggle={toggleModal}>Delete plant {plant.name} </ModalHeader>
        <ModalBody>
          Warning! This action is irreversible.
        </ModalBody>
        <ModalFooter className={uiStyles.footer}>
          <Button color="danger" onClick={confirmDeletion} data-testid="modal-button-delete">Delete!</Button>
          <Button color="secondary" onClick={toggleModal}>Cancel!</Button>
        </ModalFooter>
    </Modal>

    </>
  )
}

export default PlantsEdit
