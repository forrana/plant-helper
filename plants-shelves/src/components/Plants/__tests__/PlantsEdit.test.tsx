import React from 'react';
import { act, fireEvent, render, screen, Screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';

import { PlantType } from '../models';
import { UPDATE_PLANT } from '../queries';
import { MemoryRouter } from 'react-router-dom';
import PlantsEdit from '../PlantsEdit';
import { getRandomString } from '../utils';

const plant: PlantType = {
  id: 1,
  name: "Aloe 1",
  scientificName: "Aloe Vera" ,
  daysUntilNextWatering: 0,
  daysBetweenWatering: 7
 };

const newName = getRandomString(10)
const newScName = getRandomString(10)

const mocksWithPlant: any = [
  {
    request: {
      query: UPDATE_PLANT,
      variables: { plantId: plant.id, plantName: newName, scientificName: newScName }
    },
    result: {
      data:  plant,
    },
  },
  {
    request: {
      query: UPDATE_PLANT,
      variables: { plantName: "error", scientificName: plant.scientificName}
    },
    error:
      { message: "Something went wrong" },
  },
];

const mockCallback = jest.fn(() => null)

test('Match snapshot', async () => {
  const {container} = render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithPlant} addTypename={false}>
        <PlantsEdit plant={plant} index={0} action={mockCallback}/>
      </MockedProvider>
    </MemoryRouter>,
  );

  expect(container).toMatchSnapshot()
});

test('Allow to edit plant', async () => {
 render(
    <MemoryRouter>
      <MockedProvider mocks={mocksWithPlant} addTypename={false}>
      <PlantsEdit plant={plant} index={0} action={mockCallback}/>
      </MockedProvider>
    </MemoryRouter>,
  );

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("plant-name-input"),
      { target: { value: newName } }
    )
  });

  await act(async () => {
    fireEvent.change(
      screen.getByTestId("plant-scientific-name-input"),
      { target: { value: newScName } }
    )
  });

  await act(async () => {
    fireEvent.click(
      screen.getByTitle(/Save/i)
    )
    await new Promise(resolve => setTimeout(resolve, 0));
  });

  expect(mockCallback).toBeCalledTimes(1)
});

test('Don\'t call action on error', async () => {
    render(
       <MemoryRouter>
         <MockedProvider mocks={mocksWithPlant} addTypename={false}>
         <PlantsEdit plant={plant} index={0} action={mockCallback}/>
         </MockedProvider>
       </MemoryRouter>,
     );

     await act(async () => {
       fireEvent.change(
         screen.getByTestId("plant-name-input"),
         { target: { value: "error" } }
       )
     });

     await act(async () => {
       fireEvent.click(
         screen.getByTitle(/Save/i)
       )
       await new Promise(resolve => setTimeout(resolve, 0));
     });

     expect(mockCallback).toBeCalledTimes(0)
});

test('Don\'t update with empty value', async () => {
    render(
       <MemoryRouter>
         <MockedProvider mocks={mocksWithPlant} addTypename={false}>
         <PlantsEdit plant={plant} index={0} action={mockCallback}/>
         </MockedProvider>
       </MemoryRouter>,
     );

     await act(async () => {
       fireEvent.change(
         screen.getByTestId("plant-name-input"),
         { target: { value: newName } }
       )
     });

     await act(async () => {
       fireEvent.change(
         screen.getByTestId("plant-scientific-name-input"),
         { target: { value: "" } }
       )
     });

     await act(async () => {
       fireEvent.click(
         screen.getByTitle(/Save/i)
       )
       await new Promise(resolve => setTimeout(resolve, 0));
     });

     expect(mockCallback).toBeCalledTimes(0)
});

