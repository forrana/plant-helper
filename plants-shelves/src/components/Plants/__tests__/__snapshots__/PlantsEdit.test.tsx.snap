// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
<div>
  <form
    autocomplete="off"
    class=""
  >
    <section
      class="controls"
    >
      <button
        class="roundButton btn btn-outline-success"
        title="Save!"
        type="submit"
      >
        ✓
      </button>
      <button
        class="roundButton btn btn-outline-danger"
        title="Cancel!"
        type="button"
      >
        ❌
      </button>
    </section>
    <img
      alt="plant pot"
      class="image"
      src="pot.jpeg"
    />
    <div
      class="form-group"
    >
      <input
        class="form-control-sm form-control"
        data-testid="plant-name-input"
        id="name"
        name="name"
        placeholder="Plant name"
        required=""
        title="Plant name"
        type="text"
        value="Aloe 1"
      />
    </div>
    <div
      class="form-group"
    >
      <input
        class="form-control-sm form-control"
        data-testid="plant-scientific-name-input"
        id="scientificName"
        name="scientificName"
        placeholder="Scientific name"
        required=""
        type="text"
        value="Aloe Vera"
      />
      <ul
        class="option list-group"
      />
    </div>
    <div
      class="form-group"
    >
      <section
        class="label"
      >
        <span
          class="icon-droplet"
        />
        <h4
          class="header"
        >
          <span
            class="icon-droplet"
          />
        </h4>
      </section>
      <input
        class="form-control-range"
        data-testid="days-between-watering-input"
        id="daysBetweenWatering"
        max="30"
        min="1"
        name="daysBetweenWatering"
        placeholder="Days between watering"
        required=""
        type="range"
        value="7"
      />
      <small>
        7
         day(s)
      </small>
    </div>
  </form>
</div>
`;
